src/bin/pg_autoctl/service_keeper.c

```diff
@@ -395,6 +372,48 @@ keeper_node_active_loop(Keeper *keeper, pid_t start_pid)
 			continue;
 		}
 
+		/*
+		 * Upon receiving SIGTERM, start maintenance on the monitor, possibly
+		 * triggering a failover when the primary node is being shut down.
+		 */
+		if (asked_to_stop)
+		{
+			if (shutdownSequenceInProgress && couldStartMaintenanceAtShutdown)
+			{
+				log_trace("Shutdown in progress");
+			}
+			else if (keeperState->current_role != MAINTENANCE_STATE &&
+					 keeperState->current_role != PREPARE_MAINTENANCE_STATE)
+			{
+				int nodeId = keeper->state.current_node_id;
+				bool mayRetry = false;
+
+				log_info("Enabling maintenance on this node for shutdown");
+
+				/*
+				 * Make sure that at restart, we get out of maintenance, even
+				 * if we later receive a stronger signal such as SIGINT
+				 * (asked_to_stop_fast) or SIGQUIT (asked_to_quit).
+				 */
+				keeperState->service_state = SERVICE_SHUTDOWN_IN_MAINTENANCE;
+
+				if (keeper_store_state(keeper) &&
+					monitor_start_maintenance(monitor, nodeId, &mayRetry))
+				{
+					couldStartMaintenanceAtShutdown = true;
+				}
+				else
+				{
+					/* warning: we're going to try again */
+					log_warn("Failed to start maintenance "
+							 "upon receiving SIGTERM");
+				}
+			}
```

Comment:
Do we need another `else` here that sets `keeperState->service_state = SERVICE_SHUTDOWNING`?